File: _services/algorithm-lab/tests/exercise-lab.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import vm from 'node:vm';
const root=path.resolve(import.meta.dirname,'../../..'),lab=path.join(root,'algorithm-lab');
const problems=JSON.parse(fs.readFileSync(path.join(lab,'exercise-problems.json'),'utf8'));
let receive;
const next=()=>new Promise(resolve=>{receive=resolve;});
const ctx=vm.createContext({TextEncoder,TextDecoder,WebAssembly,Uint8Array,ArrayBuffer,DataView,console,setTimeout,clearTimeout,performance,
  fetch:async file=>{const b=fs.readFileSync(path.join(lab,String(file).split('?')[0]));return {ok:true,arrayBuffer:async()=>b.buffer.slice(b.byteOffset,b.byteOffset+b.byteLength)};},
  self:{postMessage:data=>{if(data.type!=='phase')receive(data);}}
});
ctx.importScripts=(...files)=>files.forEach(file=>vm.runInContext(fs.readFileSync(path.join(lab,file.split('?')[0]),'utf8'),ctx));
const boot=next();vm.runInContext(fs.readFileSync(path.join(lab,'c-worker.js'),'utf8'),ctx);
assert.equal((await boot).type,'ready');
async function run(p,code){const pending=next();await ctx.self.onmessage({data:{token:p.id,problem:p,cases:p.tests,code}});return (await pending).report;}
let total=0;
for(const p of problems){
  const r=await run(p,p.c.solution);
  assert.ok(!r.error,`${p.id} ${p.function}: ${r.error}`);
  assert.equal(r.passed,r.total,`${p.id}: ${JSON.stringify(r.rows.filter(x=>!x.ok))}`);
  total+=r.total;console.log(p.id,p.function,r.passed+'/'+r.total);
}
const bad={
  fibonacci:'long long fibonacci(int n){long long a=0,b=1;for(int i=0;i<n;i++){long long t=a+b;a=b;b=t;}return a+(n==90);}',
  dsu_find:'int dsu_find(int p[],int x){while(p[x]!=x)x=p[x];return x;}',
  relax:'int relax(long long d[],int u,int v,int w){return d[u]<1000000000000000000LL&&d[u]+w<d[v];}',
  fib_memo:'long long fib_memo(int n,long long m[]){long long a=0,b=1;for(int i=0;i<n;i++){long long t=a+b;a=b;b=t;}return a;}'
};
for(const [name,code] of Object.entries(bad)){
  const r=await run(problems.find(p=>p.function===name),code);
  assert.ok(r.error||r.passed<r.total,`${name}: wrong answer accepted`);
}
console.log(`${problems.length} real WebAssembly C references / ${total} cases passed; ${Object.keys(bad).length} incorrect answers rejected.`);

File: _services/algorithm-lab/tests/exercise-lab.py
"""Exercise references, actual Python worker, and wrong-answer regressions."""
import json
from pathlib import Path
ROOT=Path(__file__).resolve().parents[3]
problems=json.loads((ROOT/'algorithm-lab/exercise-problems.json').read_text(encoding='utf-8'))
grader=(ROOT/'algorithm-lab/worker.js').read_text(encoding='utf-8').split('const answer = py.runPython(`',1)[1].split('`);',1)[0]
def run(p,code):
    ns={'_payload_json':json.dumps({'problem':p,'cases':p['tests'],'code':code})}
    exec(compile(grader,'worker.py','exec'),ns)
    return ns['_answer']
total=0
for p in problems:
    r=run(p,p['solution'])
    assert not r.get('error'),(p['id'],r)
    assert r['passed']==r['total'],(p['id'],[x for x in r['rows'] if not x['ok']])
    total+=r['passed']
bad={
 'fibonacci':'def fibonacci(n):\n    a,b=0,1\n    for _ in range(n): a,b=b,a+b\n    return a+1 if n==90 else a',
 'insertion_sort':'def insertion_sort(a):\n    return sorted(a)',
 'dsu_find':'def dsu_find(parent,x):\n    while parent[x]!=x: x=parent[x]\n    return x',
 'relax':'def relax(dist,u,v,w):\n    return int(dist[u]<10**18 and dist[u]+w<dist[v])',
 'fib_memo':'def fib_memo(n,memo):\n    a,b=0,1\n    for _ in range(n): a,b=b,a+b\n    return a',
}
for name,code in bad.items():
    r=run(next(p for p in problems if p['function']==name),code)
    assert r['passed']<r['total'],(name,'wrong answer accepted')
print(f'{len(problems)} Python references / {total} cases passed; {len(bad)} incorrect answers rejected.')
